subprojects/core/src/main/java/org/gradle/internal/classpath/AccessTrackingProperties.java

```diff
@@ -173,29 +299,48 @@ public boolean containsKey(Object key) {
 
     @Override
     public Object put(Object key, Object value) {
-        return delegate.put(key, value);
+        Object oldValue;
+        synchronized (delegate) {
+            oldValue = delegate.put(key, value);
+        }
+        reportKeyAndValue(key, oldValue);
+        listener.onChange(key, value);
+        return oldValue;
+
     }
 
     @Override
     public Object setProperty(String key, String value) {
-        return delegate.setProperty(key, value);
+        return put(key, value);
     }
 
     @Override
     public Object remove(Object key) {
-        Object result = delegate.remove(key);
+        Object result;
+        synchronized (delegate) {
+            result = delegate.remove(key);
+        }
         reportKeyAndValue(key, result);
+        listener.onRemove(key);
         return result;
+
     }
 
     @Override
     public void putAll(Map<?, ?> t) {
-        delegate.putAll(t);
+        // putAll has no return value so keys do not become inputs.
+        t.forEach(listener::onChange);
```

Comment:
It isn't relevant, but I changed it just in case. FWIW, it isn't possible to have a consistent guarantee there, because set views are intentionally different to support `entrySet` modifications: its `onAccess` must happen before modification to find out original value corresponding to `entry.getKey()`. 

I documented that the state of the Properties is undefined when the listener gets called.